docs/content/python/guides/rbac.md

```diff
@@ -0,0 +1,326 @@
+
+```polar
+has_role(user: User, name: String, resource: Resource) if
+  role in user.roles and
+  role matches { name: name, resource: resource };
+```
+
+`role in user.roles` iterates over a user's assigned roles and `role matches {
+name: name, resource: resource }` succeeds if the user has been assigned the
+`name` role for `resource`.
+
+{{% callout "Note" "blue" %}}
+  The body of this rule will vary according to the way roles are stored in your
+  application. The data model for our GitClub example is as follows:
+
+  ```py
+  @dataclass(frozen=True)
+  class Organization:
+      name: str
+
+  @dataclass(frozen=True)
+  class Repository:
+      name: str
+      organization: Organization
+
+  @dataclass(frozen=True)
+  class Role:
+      name: str
+      resource: Union[Repository, Organization]
+
+  @dataclass
+  class User:
+      name: str
+      roles: Set[Role]
+
+      def assign_role_for_resource(self, name, resource):
+          self.roles.add(Role(name, resource))
+  ```
+
+  If, for example, repository roles and organization roles were stored
+  separately instead of in a heterogeneous set, we might define a pair of
+  `has_role()` rules, one for each role type:
+
+  ```polar
+  has_role(user: User, name: String, repository: Repository) if
+    role in user.repository_roles and
+    role matches { name: name, resource: resource };
+
+  has_role(user: User, name: String, organization: Organization) if
+    role in user.organization_roles and
+    role matches { name: name, resource: resource };
+  ```
+
+{{% /callout %}}
+
+Our `has_role()` rule can check role assignments on repositories and
+organizations, but so far we've only talked about repository roles. Let's
+change that and see how Oso can leverage parent-child relationships like the
+one between `Repository` and `Organization` to grant a role on a child resource
+to a role on the parent.
+
+<!-- TODO(gj): better heading -->
+## Grant a role on a child resource to a role on the child's parent
+
+If you've used ~GitHub~ *GitClub* before, you know that having a role on an
+organization grants certain roles and permissions on that organization's
+repositories. For example, a user is granted the `"maintainer"` role on a
+repository if they're assigned the `"owner"` role on the repository's parent
+organization. This is how you write that rule with Oso:
+
+{{< code file="rbac.polar" >}}
+resource Repository {
+  # ...
+
+  relations = { parent: Organization };
+
+  "maintainer" if "owner" on "parent";
+}
+
+has_relation(organization: Organization, "parent", repository: Repository) if
+  organization = repository.organization;
+{{< /code >}}
+
+First, we declare that the `"parent"` relation for a `Repository` is an
+`Organization`:
+
+```polar
+relations = { parent: Organization };
+```
+
+This is a dictionary where each key is the name of the relation and each value
+is the relation's type.
+
+Next, we write a `has_relation()` rule that tells Oso how to check if an
```

Comment:
This ordering of arguments has confused me several times because you write the `relation` declaration on `Repository`, which makes me think `Repository` should be the first argument (not sure why, I guess because `self` usually goes in the first position -- even though this isn't a method maybe that's the analogy in my head)

I wonder if we can add some callout explaining why this is consistent with the way `has_role` and other rules work.

Reply:
94f34c9c